File: cpp/pascals_triangle.cc
/*
 * Filename: pascals_triangle.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 04 Aug 2015 07:28:51 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		vector<vector<int> > generate(int numRows)
		{
			vector<vector<int> > res;
			generate_aux(numRows, res);
			return res;
		}

	private:
		void generate_aux(
				int n,
				vector<vector<int> >& res)
		{
			if (!n) {
				return;
			}

			if (n == 1) {
				vector<int> v (1,1);
				res.push_back(v);
				return;
			}

			generate_aux(n-1, res);
			vector<int> last = res.back();
			vector<int> v (last.size() + 1, 1);

			for (int i = 1; i < last.size(); i++)
			{
				v[i] = last[i] + last[i-1];
			}

			res.push_back(v);
		}
};

File: java/PascalsTriangle.java
/*
 * Filename: PascalsTriangle.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 19 Jun 2015 01:13:39 PM CST
 */

public class Solution {
	public List<List<Integer>> generate(int numRows) {
		if (numRows <= 0) {
			return new ArrayList<List<Integer>>();
		}
		
		if (numRows == 1) {
			List<List<Integer>> res = new ArrayList<List<Integer>>();
			List<Integer> row = new ArrayList<Integer>();
			row.add(1);
			res.add(row);
			return res;
		}

		List<List<Integer>> recRes = generate(numRows - 1);
		List<Integer> lastRow = recRes.get(recRes.size() - 1);
		List<Integer> row = new ArrayList<Integer>();

		for (int i = 0; i < lastRow.size(); i++) {
			if (i == 0) {
				row.add(1);
			} else {
				row.add(lastRow.get(i) + lastRow.get(i-1));
			}
		}

		row.add(1);
		recRes.add(row);
		return recRes;
	}
}
